File: js/popnei/bench/time_pop_dists.mjs
/**
 * How long the distances between populations take in WebAssembly under node,
 * and how long reading the same file alone takes.
 *
 * Nothing had timed `calcPopDists` in wasm when this was written. "Speed" of
 * docs/specs/dists.md names the dataset it is to be timed on, 100000
 * variants x 1000 individuals, biallelic, 3 in 100 genotypes missing, read
 * from a vars file, with the individuals cut into 3 populations and into 20,
 * so that how the cost grows with the pairs is measured and not guessed: 3
 * populations make 3 pairs and 20 make 190. That section has no number to
 * reach yet, and this is the clock in wasm that gives the first one.
 * `crates/popnei/benches/time_pop_dists.py` is the same clock on the native
 * build.
 *
 * It times two passes over one source. The calculation is one call of
 * `calcPopDists`, which reads every block and gives the seven measures of
 * every pair of populations with the standard error of each beside it. The
 * reading alone is `iterBlocks({fields: ["chrom", "pos"]})`, the same pass
 * carrying the two fields that calculation asks its reader for, with nothing
 * done to the genotypes but adding up how many alleles came out. The
 * difference of the two is what the calculation costs beyond the reader,
 * which is the number "Speed" asks for.
 *
 * The standard errors are resampled over groups of variants, each group a
 * stretch of one chromosome, and the fourth argument is how long one stretch
 * is in base pairs, 1000000 when it is not given. The dataset above is two
 * chromosomes of 50000 variants 1000 base pairs apart, which 1000000 cuts
 * into 100 groups. popnei refuses a pass that falls into fewer than 20
 * groups, so a shorter file needs a shorter stretch: the 5000 variants of
 * `big5000.vars`, 1000 base pairs apart, fall into 5 groups with 1000000,
 * which is refused, and into 25 with 200000.
 *
 * A populations file holds one line for each individual, the name of the
 * individual and the name of its population with a tab between them, which
 * is what `crates/popnei/benches/make_pops.py` writes. The populations are
 * given to `calcPopDists` in the order in which they first appear in the
 * file, and the pairs are in that order too, so a population whose name is a
 * whole number must not be used: JavaScript keeps such a name at the front
 * of an object whatever order it was put in.
 *
 * Reading the vars file into a `Uint8Array` and `openVars` are outside both
 * timings, as they are outside the native ones: `openVars` copies those
 * bytes into the memory of wasm and reads the schema and the footer, and
 * every pass reads them again from their start. Reading the populations file
 * is outside both as well, and is node's own work. What the two reads and
 * `openVars` took is printed under the runs. wasm has no threads, so there
 * is one.
 *
 *     node js/popnei/bench/time_pop_dists.mjs <path to a vars file>
 *         <path to a populations file> <runs> [group length in base pairs]
 *
 * It needs `npm run build` in `js/popnei` before it: that is what compiles
 * the core to WebAssembly in release and writes `wasm/` and `dist/`, which
 * the name `popnei` resolves to here. One pass of each kind runs before the
 * timed ones and is not timed, and the two kinds are run one after the other
 * inside each round, so that a machine that grows busier over the runs falls
 * on both.
 *
 * Beside each timed calculation it prints the variants the pass took, the
 * pairs of the result, how many resampling groups the variants fell into and
 * Hudson's F_ST of the first pair with its standard error, which together
 * show that the call read the genotypes and resampled them. A call whose
 * pairs are not the pairs the populations of the file make is refused with a
 * message, because a run over the wrong pairs is a time nobody can use.
 */

import { readFile } from "node:fs/promises";

import { calcPopDists, init, openVars } from "popnei";

/** How long one resampling group is when the command line does not say it,
 * which cuts the dataset of "Speed" into 100 groups. */
const DEFAULT_GROUP_BASE_PAIRS = 1000000;

const USAGE =
  "node js/popnei/bench/time_pop_dists.mjs <path to a vars file> " +
  "<path to a populations file> <runs> [group length in base pairs]";

/**
 * The populations of a populations file, the name of each population to the
 * names of its individuals, in the order the populations first appear.
 *
 * @throws {Error} When a line is not the name of an individual, a tab and
 * the name of its population, with the file and the line in the message.
 */
function popsOfTheFile(text, path) {
  const ofEachPop = new Map();
  for (const [which, line] of text.split(/\r?\n/).entries()) {
    if (line === "") {
      continue;
    }
    const fields = line.split("\t");
    if (fields.length !== 2 || fields[0] === "" || fields[1] === "") {
      throw new Error(
        `${path}, line ${which + 1}: a line of a populations file is the ` +
          "name of an individual, a tab and the name of its population, " +
          `and this one is \`${line}\``,
      );
    }
    const [individual, pop] = fields;
    const ofThePop = ofEachPop.get(pop);
    if (ofThePop === undefined) {
      ofEachPop.set(pop, [individual]);
    } else {
      ofThePop.push(individual);
    }
  }
  if (ofEachPop.size === 0) {
    throw new Error(`${path}: the populations file holds no individual`);
  }
  return Object.fromEntries(ofEachPop);
}

/**
 * One call of `calcPopDists`, as what it gave and the variants it took.
 *
 * What it gave is the line printed beside the time: the pairs, the
 * resampling groups, and Hudson's F_ST of the first pair with its standard
 * error.
 *
 * @throws {Error} When the pairs of the result are not `numPairsOfThePops`,
 * the pairs the populations make, and when the result carries no F_ST, no
 * standard error or no value for the first pair.
 */
function theCalculation(variants, pops, groupBasePairs, numPairsOfThePops) {
  const popDists = calcPopDists(variants, pops, {
    jackknifeGroup: groupBasePairs,
  });
  if (popDists.numPairs !== numPairsOfThePops) {
    throw new Error(
      `the ${Object.keys(pops).length} populations of the file make ` +
        `${numPairsOfThePops} pairs, and the call gave ${popDists.numPairs}: ` +
        "the run is not over the pairs it was asked for",
    );
  }
  const fst = popDists.fst;
  if (fst === null || fst.standardErrors === null) {
    throw new Error(
      "the call gave no F_ST, or gave it with no standard error, so it did " +
        "not do the work this times",
    );
  }
  const value = fst.distVector[0];
  const standardError = fst.standardErrors[0];
  if (value === undefined || standardError === undefined) {
    throw new Error("the call gave no value for the first pair");
  }
  const [firstPop, secondPop] = popDists.pops;
  return [
    `${popDists.numPairs} pairs, ${popDists.numGroups} resampling groups, ` +
      `fst of ${firstPop} and ${secondPop} ${value.toFixed(6)} ` +
      `+- ${standardError.toFixed(6)}`,
    popDists.passStats.numVars,
  ];
}

/**
 * One pass carrying the chromosome and the position, which are the fields
 * `calcPopDists` asks its reader for when the groups are stretches of base
 * pairs, as the alleles that came out of it and the variants it took.
 *
 * The alleles are added up so that a reader which stopped filling the
 * genotypes, or filled them only when somebody read them, would show.
 */
function theReadingAlone(variants) {
  const blocks = variants.iterBlocks({ fields: ["chrom", "pos"] });
  let alleles = 0;
  for (const block of blocks) {
    alleles += block.gts.length;
  }
  return [`${alleles} alleles`, blocks.passStats.numVars];
}

/**
 * The same pass with nothing read out of the block at all, which is the
 * baseline to subtract.
 *
 * `theReadingAlone` above reads `block.gts`, and a column of a block in
 * wasm is copied when it crosses: over the dataset of "Speed" that is
 * 200 million bytes into fresh typed arrays, and reaching the block at all
 * builds a string for the chromosome of every variant, 100000 of them.
 * `calcPopDists` copies none of that: no block of it leaves the core, and
 * it makes one string for each resampling group, 100 of them. So
 * subtracting `theReadingAlone` takes out more than the reader and leaves
 * the calculation looking faster than it is. This pass reads only what the
 * reader counted, so the bytes stay in wasm, and it is the honest
 * subtrahend. The performance review of 23 September 2026 found this.
 */
function theReadingWithNothingRead(variants) {
  const blocks = variants.iterBlocks({ fields: ["chrom", "pos"] });
  let numBlocks = 0;
  for (const _block of blocks) {
    numBlocks += 1;
  }
  return [`${numBlocks} blocks`, blocks.passStats.numVars];
}

/** The best, the median and the worst of `times`, in seconds. */
function saidAbout(what, times) {
  const sorted = [...times].sort((one, other) => one - other);
  const median = sorted[(sorted.length - 1) >> 1];
  return (
    `${what}: best ${sorted[0].toFixed(3)} s, ` +
    `median ${median.toFixed(3)} s, ` +
    `worst ${sorted[sorted.length - 1].toFixed(3)} s`
  );
}

const [path, popsPath, runsAsText, groupAsText] = process.argv.slice(2);
if (path === undefined || popsPath === undefined || runsAsText === undefined) {
  console.log(USAGE);
  process.exit(1);
}
const runs = Number.parseInt(runsAsText, 10);
const groupBasePairs =
  groupAsText === undefined
    ? DEFAULT_GROUP_BASE_PAIRS
    : Number.parseInt(groupAsText, 10);
if (!Number.isInteger(runs) || runs < 1) {
  console.log(`the runs are a whole number of 1 or more\n${USAGE}`);
  process.exit(1);
}
if (!Number.isInteger(groupBasePairs) || groupBasePairs < 1) {
  console.log(
    `the length of a resampling group is a whole number of base pairs of 1 ` +
      `or more\n${USAGE}`,
  );
  process.exit(1);
}

let started = performance.now();
const pops = popsOfTheFile(await readFile(popsPath, "utf8"), popsPath);
const popsReadIn = (performance.now() - started) / 1000;
const numPops = Object.keys(pops).length;
const numPairsOfThePops = (numPops * (numPops - 1)) / 2;
console.log(
  `${path}, the ${numPops} populations of ${popsPath}, ` +
    `${numPairsOfThePops} pairs, resampling groups of ${groupBasePairs} ` +
    `base pairs, popnei in wasm under node, ${runs} runs, one thread`,
);

await init();
started = performance.now();
const bytes = new Uint8Array(await readFile(path));
const readIn = (performance.now() - started) / 1000;
started = performance.now();
const variants = openVars(bytes);
const openedIn = (performance.now() - started) / 1000;

console.log(
  "the first run of each kind, which is not timed: " +
    `${theCalculation(variants, pops, groupBasePairs, numPairsOfThePops)}`,
);
theReadingAlone(variants);
theReadingWithNothingRead(variants);

const calculations = [];
const readings = [];
const bareReadings = [];
for (let run = 1; run <= runs; run += 1) {
  for (const [what, passOfTheRun, times] of [
    [
      "the calculation",
      () => theCalculation(variants, pops, groupBasePairs, numPairsOfThePops),
      calculations,
    ],
    ["the reading alone", () => theReadingAlone(variants), readings],
    [
      "the reading with nothing read",
      () => theReadingWithNothingRead(variants),
      bareReadings,
    ],
  ]) {
    started = performance.now();
    const [counted, numVars] = passOfTheRun();
    const took = (performance.now() - started) / 1000;
    times.push(took);
    console.log(
      `run ${run}, ${what}: ${took.toFixed(3)} s, ${numVars} variants, ${counted}`,
    );
  }
}
console.log(saidAbout("the calculation", calculations));
console.log(saidAbout("the reading alone", readings));
console.log(saidAbout("the reading with nothing read", bareReadings));
const difference = Math.min(...calculations) - Math.min(...bareReadings);
console.log(
  `the calculation with the reading taken out, on the bests: ` +
    `${difference.toFixed(3)} s, which subtracts the reading with nothing ` +
    `read: subtracting the reading alone instead would give ` +
    `${(Math.min(...calculations) - Math.min(...readings)).toFixed(3)} s, ` +
    `and that one takes out the copies out of wasm as well as the reader`,
);
console.log(
  `reading the vars file into a Uint8Array, which is in neither: ` +
    `${readIn.toFixed(3)} s; openVars, which is in neither: ` +
    `${openedIn.toFixed(3)} s; reading the populations file, which is in ` +
    `neither: ${popsReadIn.toFixed(3)} s`,
);
variants.free();

File: crates/popnei/benches/time_pop_dists.py
"""How long the distances between populations take over a whole vars file,
and how long reading that file alone takes.

Nothing had timed `calc_pop_dists` when this file was written. "Speed" of
docs/specs/dists.md asks for it on 100000 variants x 1000 individuals,
biallelic, 3 in 100 genotypes missing, read from a vars file, with the
individuals cut into 3 populations and into 20, and with the time of reading
that file measured beside the calculation and taken out, so that the reader
is not in the number. `time_pop_dists_pynei.py`, beside this file, is the
same clock on pyNei's `calc_jost_dest_pop_dists`, which is the program the
spec compares with. `time_kosman_dists.py` is the same clock on the
distances between individuals, and this file has its shape with another call
in it.

It times two passes over the same source. The **calculation** is one call of
`calc_pop_dists`, which reads every block of the source and gives the seven
measures of how far apart two populations are for every pair of them, each
value with the standard error of a block jackknife beside it. The **reading
alone** is `iter_blocks(fields=("chrom", "pos"))`, the same pass carrying the
genotypes, the chromosome and the position, which are the fields the
calculation asks its reader for when the groups are stretches of a
chromosome, with nothing done to them but adding up how many alleles and how
many positions each block says it holds, which is their shape and not their
content. The difference of the two is what the calculation costs beyond the
reader, and it is the number the spec asks for.

The populations come from a file of one line for each individual: the name
of the individual, a tab, and the name of its population. `make_pops.py`,
beside this file, wrote the two the spec asks for, `pops3.tsv` of 3
populations and `pops20.tsv` of 20. The populations are kept in the order in
which the file first names each of them, and that is the order of the pairs
of the result.

The fourth argument is the length in base pairs of a resampling group, the
stretch of one chromosome that the standard errors are resampled over, and
it is 1000000 when it is not given. `calc_pop_dists` has no default for it:
its `jackknife_group` is a required argument, because a length is a claim
about how far linkage reaches in the dataset at hand and popnei does not
make that claim for a user. On the dataset of the spec, two chromosomes of
50000 variants 1000 base pairs apart, 1000000 cuts the variants into 100
groups. A pass whose variants fall into fewer than 20 groups is a
`ValueError`, so a shorter file needs a shorter group: 5000 variants of that
spacing fall into 5 groups at 1000000 and into 50 at 100000.

All seven measures are calculated, which is what leaving `measures` out
asks for, and `min_num_individuals` is left at its default of 20 called
genotypes.

`open_vcf` and `open_vars` are outside both timings. They read the header of
the VCF, or the schema and the footer of the vars file, and no genotype:
every pass reads the source again from its start, so the handle is built
once and given to every run. What the open itself took is printed under the
runs.

The threads are rayon's, which the library never sets: the pool takes one
thread for each core unless `RAYON_NUM_THREADS` is in the environment of the
process before the first call, which is why it is set in the command and not
from Python. Over a vars file the reader runs on the thread that calls it,
so there the threads are the calculation's alone.

    RAYON_NUM_THREADS=1 uv run python time_pop_dists.py <path> <populations file> <runs>
    uv run python time_pop_dists.py <path> <populations file> <runs> <group length>

A path that ends in `.vars` is opened with `open_vars` and anything else
with `open_vcf`. One pass of each kind runs before the timed ones and is not
timed, for the page cache. The two kinds are run one after the other inside
each round, so that a machine that grows busier over the runs falls on both.

popnei has to be built in release, `uv run maturin develop --release`. What
`uv run maturin develop` makes is a debug build of the core, and for the
Kosman distances over the same file it was about fifty times slower, 65.9 s
against 1.275 s on one thread; nobody has measured what that build costs
here, and no number taken with it says anything about the calculation.
"""

import os
import statistics
import sys
import time

from popnei import calc_pop_dists, open_vars, open_vcf

# The length in base pairs of a resampling group when the command gives
# none. It cuts the 100000 variants of the dataset of "Speed" of
# docs/specs/dists.md, two chromosomes of variants 1000 base pairs apart,
# into 100 groups.
GROUP_LENGTH = 1_000_000


def the_pops(path: str) -> dict[str, list[str]]:
    """The populations of the file at `path`, as the dict of population name
    to the names of its individuals that `calc_pop_dists` takes.

    The file holds one line for each individual, its name, a tab and the name
    of its population. The populations come out in the order in which the
    file first names each of them, and the individuals of each in the order
    of the file.
    """
    pops: dict[str, list[str]] = {}
    with open(path) as fhand:
        for number, line in enumerate(fhand, start=1):
            if not line.strip():
                continue
            written = line.rstrip("\n")
            individual, tab, pop = written.partition("\t")
            if not tab or not individual or not pop:
                raise ValueError(
                    f"{path}, line {number}: a line of a populations file is "
                    f"the name of an individual, a tab and the name of its "
                    f"population, and this one is {written!r}"
                )
            pops.setdefault(pop, []).append(individual)
    return pops


def the_calculation(variants, pops, group_length: int) -> tuple[str, int]:
    """One call of `calc_pop_dists` over `variants` for the pairs of `pops`,
    with the standard errors resampled over groups of `group_length` base
    pairs, as what the result holds and the variants the pass took.

    The value printed is Hudson's F_ST of the first pair with its standard
    error, one of the seven measures the call gives, so that a run whose
    pass read nothing shows itself.
    """
    dists = calc_pop_dists(variants, pops, jackknife_group=group_length)
    fst = dists.fst
    said = (
        f"{fst.dist_vector.shape[0]} pairs, {len(dists.group_ids)} groups, "
        f"F_ST of {dists.pops[0]} and {dists.pops[1]} "
        f"{fst.dist_vector[0]:.6f} +- {fst.standard_errors[0]:.6f} over "
        f"{int(dists.num_vars[0])} variants"
    )
    return said, dists.pass_stats.num_vars


def the_reading_alone(variants) -> tuple[str, int]:
    """One pass over `variants` carrying the genotypes, the chromosome and
    the position, as the alleles and the positions its blocks say they hold
    and the variants the pass took.

    `block.gts.size` and `block.pos.size` are the shapes of the arrays the
    core filled, which reach numpy without a copy, so adding them up shows a
    pass whose blocks carry neither column and does not show a buffer of the
    right shape that nothing wrote into. Nothing here reads a genotype or a
    position.
    """
    blocks = variants.iter_blocks(fields=("chrom", "pos"))
    alleles = 0
    positions = 0
    for block in blocks:
        alleles += int(block.gts.size)
        positions += int(block.pos.size)
    return f"{alleles} alleles, {positions} positions", blocks.pass_stats.num_vars


def said_about(what: str, times: list[float]) -> str:
    """The best, the median and the worst of `times`, under the name of the
    pass they are of."""
    return (
        f"{what}: best {min(times):.3f} s, median {statistics.median(times):.3f} s, "
        f"worst {max(times):.3f} s"
    )


def main() -> int:
    if len(sys.argv) not in (4, 5):
        print(__doc__)
        return 1
    path = sys.argv[1]
    pops_path = sys.argv[2]
    runs = int(sys.argv[3])
    group_length = int(sys.argv[4]) if len(sys.argv) == 5 else GROUP_LENGTH
    pops = the_pops(pops_path)
    threads = os.environ.get("RAYON_NUM_THREADS", "one for each core")
    print(
        f"{path}, popnei, {len(pops)} populations of {pops_path}, {runs} runs, "
        f"jackknife_group={group_length}, RAYON_NUM_THREADS={threads}"
    )

    started = time.perf_counter()
    variants = open_vars(path) if path.endswith(".vars") else open_vcf(path)
    opened_in = time.perf_counter() - started

    print(
        f"the first run of each kind, which is not timed: "
        f"{the_calculation(variants, pops, group_length)}"
    )
    the_reading_alone(variants)

    calculations: list[float] = []
    readings: list[float] = []
    for run in range(1, runs + 1):
        for what, pass_of_the_run, times in (
            (
                "the calculation",
                lambda variants: the_calculation(variants, pops, group_length),
                calculations,
            ),
            ("the reading alone", the_reading_alone, readings),
        ):
            started = time.perf_counter()
            counted, num_vars = pass_of_the_run(variants)
            took = time.perf_counter() - started
            times.append(took)
            print(f"run {run}, {what}: {took:.3f} s, {num_vars} variants, {counted}")
    print(said_about("the calculation", calculations))
    print(said_about("the reading alone", readings))
    print(
        f"the calculation with the reading taken out, on the bests: "
        f"{min(calculations) - min(readings):.3f} s"
    )
    print(f"opening the source, which is in neither: {opened_in:.4f} s")
    return 0


if __name__ == "__main__":
    sys.exit(main())
